### ISSUE TITLE
Cannot pull from Private Docker Registry

### ISSUE THREAD

**Issue by lorenzvth7** [docker/docker#18742]
```
docker version
Client:
 Version:      1.8.2
 API version:  1.20
 Package Version: docker-1.8.2-7.el7.centos.x86_64
 Go version:   go1.4.2
 Git commit:   bb472f0/1.8.2
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.2
 API version:  1.20
 Package Version: 
 Go version:   go1.4.2
 Git commit:   bb472f0/1.8.2
 Built:        
 OS/Arch:      linux/amd64
```

I have a cluster with one master and two nodes (node1 and node2). The nodes can contain docker containers.
I've my Private Docker Regsitry on Centos7 (EC2 instance) on node1 and I'm able to push to this registry.
This is possible from node1 AND node2. So every push is possible. After pushing I can see the images in my Docker repository container in:
```
root@e65daf20372e:/var/lib/registry/docker/registry/v2/repositories# ls
busybox  ubuntu
```
My Private Registry is using selfsigned certificates. 
But now the issue:

I'm not able to pull the existing images from my registry (not on node1, not on node2), it's trying to pull for a long time, after that the error shows up:
```
Trying to pull repository myrepo:5000/busybox ... not found                   
Error: image busybox:latest not found
```

The logs of my container are showing:
```

10.0.0.xx - - [17/Dec/2015:08:00:26 +0000] "PUT /v2/busybox/manifests/latest HTTP/1.1" 201 0 "" "docker/1.8.2 go/go1.4.2 kernel/3.10.0-229.20.1.el7.x86_64 os/linux arch/amd64"
time="2015-12-17T08:01:35Z" level=info msg="response completed" go.version=go1.5.2 http.request.host="myrepo:5000" http.request.id=94039036-e514-4955-adb9-7112a9d36970 http.request.method=GET http.request.remoteaddr="10.0.0.xx:60314" http.request.uri="/v2/" http.request.useragent="docker/1.8.2 go/go1.4.2 kernel/3.10.0-229.20.1.el7.x86_64 os/linux arch/amd64" http.response.contenttype="application/json; charset=utf-8" http.response.duration=3.466144ms http.response.status=200 http.response.written=2 instance.id=c6a37849-2cca-4ab5-9d6d-645ad2a44795 version=v2.2.1 
10.0.0.xx - - [17/Dec/2015:08:01:35 +0000] "GET /v2/ HTTP/1.1" 200 2 "" "docker/1.8.2 go/go1.4.2 kernel/3.10.0-229.20.1.el7.x86_64 os/linux arch/amd64"
time="2015-12-17T08:01:35Z" level=info msg="response completed" go.version=go1.5.2 http.request.host="myrepo:5000" http.request.id=5acdc8b2-d92c-4be7-939b-139012419a07 http.request.method=GET http.request.remoteaddr="10.0.0.xx:60324" http.request.uri="/v2/busybox/manifests/latest" http.request.useragent="docker/1.8.2 go/go1.4.2 kernel/3.10.0-229.20.1.el7.x86_64 os/linux arch/amd64" http.response.contenttype="application/json; charset=utf-8" http.response.duration=4.165926ms http.response.status=200 http.response.written=3851 instance.id=c6a37849-2cca-4ab5-9d6d-645ad2a44795 version=v2.2.1 
10.0.0.xx - - [17/Dec/2015:08:01:35 +0000] "GET /v2/busybox/manifests/latest HTTP/1.1" 200 3851 "" "docker/1.8.2 go/go1.4.2 kernel/3.10.0-229.20.1.el7.x86_64 os/linux arch/amd64"
time="2015-12-17T08:17:07Z" level=info msg="response completed" go.version=go1.5.2 http.request.host="myrepo:5000" http.request.id=2f8248b3-3bdd-47cb-a80b-8fa6f789e72e http.request.method=GET http.request.remoteaddr="10.0.0.xx:39317" http.request.uri="/v2/" http.request.useragent="docker/1.8.2 go/go1.4.2 kernel/3.10.0-229.20.1.el7.x86_64 os/linux arch/amd64" http.response.contenttype="application/json; charset=utf-8" http.response.duration=2.454634ms http.response.status=200 http.response.written=2 instance.id=c6a37849-2cca-4ab5-9d6d-645ad2a44795 version=v2.2.1 
10.0.0.xx - - [17/Dec/2015:08:17:07 +0000] "GET /v2/ HTTP/1.1" 200 2 "" "docker/1.8.2 go/go1.4.2 kernel/3.10.0-229.20.1.el7.x86_64 os/linux arch/amd64"
10.0.0.xx - - [17/Dec/2015:08:17:07 +0000] "GET /v1/repositories/busybox/images HTTP/1.1" 404 19 "" "docker/1.8.2 go/go1.4.2 kernel/3.10.0-229.20.1.el7.x86_64 os/linux arch/amd64"
```
I also got a ```"GET /v1/repositories/ubuntu/images HTTP/1.1" 404 19 ""``` while my image is in the path: ```/var/lib/registry/docker/registry/v2/repositories# ls
busybox  ubuntu```
so It's using V2

**Comment by dmcgowan** [comment]
Can you send the daemon debugs logs as well. It is unclear why the pull operation did a fallback to v1. Add the `-D` flag to the daemon process to get the debug logs.

**Comment by lorenzvth7** [comment]
When I perform the ```Docker Daemon -D``` command I got this error:

```ERRO[0000] WARNING: No --storage-opt dm.thinpooldev specified, using loopback; this configuration is strongly discouraged for production use 
DEBU[0000] activateDeviceIfNeeded()                     
DEBU[0000] Error device setupBaseImage: Base Device UUID verification failed. Possibly using a different thin pool then last invocation:Error running DeviceCreate (ActivateDevice) dm_task_run failed 
ERRO[0000] [graphdriver] prior storage driver "devicemapper" failed: Base Device UUID verification failed. Possibly using a different thin pool then last invocation:Error running DeviceCreate (ActivateDevice) dm_task_run failed FATA[0000] Error starting daemon: error initializing graphdriver: Base Device UUID verification failed. Possibly using a different thin pool then last invocation:Error running DeviceCreate (ActivateDevice) dm_task_run failed```

Info:
```
[root@ip-10-0-0-15 centos]# pvscan
  PV /dev/xvdb1   VG docker-vg   lvm2 [40.00 GiB / 20.73 GiB free]
  Total: 1 [40.00 GiB] / in use: 1 [40.00 GiB] / in no VG: 0 [0   ]
```

To install Docker I followed the prerequisitions of OpenShift:
https://docs.openshift.org/latest/getting_started/administrators.html#prerequisites

**Comment by lorenzvth7** [comment]
I have posted the logs which contain the container ID of my registry (ae8fc4363b02):
```
DEBU[0000] killing old running container ae8fc4363b02163ae3b05a76de1e33ae83537328d29d3a6c05fbaef2049e87fd 
DEBU[0000] [devmapper] UnmountDevice(hash=ae8fc4363b02163ae3b05a76de1e33ae83537328d29d3a6c05fbaef2049e87fd) 
DEBU[0000] [devmapper] UnmountDevice(hash=ae8fc4363b02163ae3b05a76de1e33ae83537328d29d3a6c05fbaef2049e87fd) END 
ERRO[0000] Error unmounting device ae8fc4363b02163ae3b05a76de1e33ae83537328d29d3a6c05fbaef2049e87fd: UnmountDevice: device not-mounted id ae8fc4363b02163ae3b05a76de1e33ae83537328d29d3a6c05fbaef2049e87fd

DEBU[0000] Starting container ae8fc4363b02163ae3b05a76de1e33ae83537328d29d3a6c05fbaef2049e87fd 
DEBU[0000] activateDeviceIfNeeded(ae8fc4363b02163ae3b05a76de1e33ae83537328d29d3a6c05fbaef2049e87fd) 
```
```
DEBU[0000] /sbin/iptables, [--wait -t nat -A DOCKER -p tcp -d 0/0 --dport 5000 -j DNAT --to-destination 172.17.0.1:5000 ! -i docker0] 
DEBU[0000] /sbin/iptables, [--wait -t filter -A DOCKER ! -i docker0 -o docker0 -p tcp -d 172.17.0.1 --dport 5000 -j ACCEPT] 
DEBU[0000] /sbin/iptables, [--wait -t nat -A POSTROUTING -p tcp -s 172.17.0.1 -d 172.17.0.1 --dport 5000 -j MASQUERADE] 
DEBU[0000] /sbin/iptables, [--wait -t nat -D DOCKER -p tcp -d 0/0 --dport 5000 -j DNAT --to-destination 172.17.0.1:5000 ! -i docker0] 
DEBU[0000] /sbin/iptables, [--wait -t filter -D DOCKER ! -i docker0 -o docker0 -p tcp -d 172.17.0.1 --dport 5000 -j ACCEPT] 
DEBU[0000] /sbin/iptables, [--wait -t nat -D POSTROUTING -p tcp -s 172.17.0.1 -d 172.17.0.1 --dport 5000 -j MASQUERADE] 
WARN[0000] Failed to allocate and map port 5000: Error starting userland proxy: listen tcp 0.0.0.0:5000: bind: address already in use 
ERRO[0000] error locating network id 0f011e1e5501e06f3b393918bdccb513d718e9951ac6d7784d571c322657da6a: network 0f011e1e5501e06f3b393918bdccb513d718e9951ac6d7784d571c322657da6a not found 
DEBU[0000] [devmapper] UnmountDevice(hash=ae8fc4363b02163ae3b05a76de1e33ae83537328d29d3a6c05fbaef2049e87fd) 
DEBU[0000] [devmapper] Unmount(/var/lib/docker/devicemapper/mnt/ae8fc4363b02163ae3b05a76de1e33ae83537328d29d3a6c05fbaef2049e87fd) 
DEBU[0000] [devmapper] Unmount done                     
DEBU[0000] [devmapper] deactivateDevice(ae8fc4363b02163ae3b05a76de1e33ae83537328d29d3a6c05fbaef2049e87fd) 
DEBU[0000] [devmapper] removeDevice START(docker-202:1-16828736-ae8fc4363b02163ae3b05a76de1e33ae83537328d29d3a6c05fbaef2049e87fd) 
DEBU[0020] [devmapper] removeDevice END(docker-202:1-16828736-ae8fc4363b02163ae3b05a76de1e33ae83537328d29d3a6c05fbaef2049e87fd) 
DEBU[0020] [devmapper] deactivateDevice END(ae8fc4363b02163ae3b05a76de1e33ae83537328d29d3a6c05fbaef2049e87fd) 
DEBU[0020] [devmapper] UnmountDevice(hash=ae8fc4363b02163ae3b05a76de1e33ae83537328d29d3a6c05fbaef2049e87fd) END 
ERRO[0020] Error unmounting device ae8fc4363b02163ae3b05a76de1e33ae83537328d29d3a6c05fbaef2049e87fd: Device is Busy 
DEBU[0020] Failed to start container ae8fc4363b02163ae3b05a76de1e33ae83537328d29d3a6c05fbaef2049e87fd: Error starting userland proxy: listen tcp 0.0.0.0:5000: bind: address already in use 
INFO[0020] Daemon has completed initialization          
```

**Comment by dmcgowan** [comment]
If you are having issues getting the openshift registry up and running it is probably better to file an issue over there.  If you get the registry back up and running correctly and still having issues, can you send the section of the daemon log related to the pull operation. If you determine the issue to be your registry setup please close the issue. Thanks and feel free to reach out on IRC if you are having trouble with your setup.

**Comment by mrwilby** [comment]
I have something similar - client (of the private docker registry) is running docker 1.9.1 on Amazon Linux:

>docker info
Containers: 1
Images: 6
Server Version: 1.9.1
Storage Driver: devicemapper
 Pool Name: docker-202:1-131386-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: 
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 318.6 MB
 Data Space Total: 107.4 GB
 Data Space Available: 30.31 GB
 Metadata Space Used: 921.6 kB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.147 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.93-RHEL7 (2015-01-28)
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.13-19.30.amzn1.x86_64
Operating System: Amazon Linux AMI 2015.09
CPUs: 1
Total Memory: 1.957 GiB
Name: ip-172-16-3-128
ID: T5MP:QJJC:5C5H:FB7M:V35J:OFMR:XKLP:HSML:QZVA:T5TP:IVLI:IPVH

And the registry is running 1.7.1:

>docker --version:
Docker version 1.7.1, build 786b29d/1.7.1

>docker info:
Containers: 3
Images: 42
Storage Driver: devicemapper
 Pool Name: docker-202:1-131374-pool
 Pool Blocksize: 65.54 kB
 Backing Filesystem: extfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 1.129 GB
 Data Space Total: 107.4 GB
 Data Space Available: 29.52 GB
 Metadata Space Used: 2.273 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.145 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.93-RHEL7 (2015-01-28)
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.10-17.31.amzn1.x86_64
Operating System: Amazon Linux AMI 2015.09
CPUs: 1
Total Memory: 1.957 GiB
Name: ip-172-17-3-124
ID: GVI4:KCTG:EW5A:IOH3:SO7O:4FCY:YMCT:6J6S:QJK3:UPH6:DND4:KXUT

I see a bunch of HEAD requests for blobs using a v2 base URL (http.response.status 307) and then two GET requests on v1 endpoint which 404's.

Switching the client back to 1.7.1 solves the problem.

**Comment by lorenzvth7** [comment]
@dmcgowan the issue is not with my openshift registry (that is working fine, secure and using nfs). It's just a pure docker registry on a server which is a node of my openshift cluster. But the registry is completely seperate from it. 

What I've done:
- generate certificates
- create registry using the certificates
- copy the certificates to /etc/docker/certs.d/registryname:5000

But for every push I get a TLS handshake timeout
